File: apps/desktop/src-tauri/src/downloads/trickplay.rs
//! Tuiles trickplay (aperçu au survol de la barre) téléchargées localement.
//!
//! Manifeste : `item.Trickplay[mediaSourceId][width]` = TrickplayInfo (requiert
//! `fields=Trickplay` sur l'item). Nombre de planches (vérifié source Jellyfin
//! v10.11) : `ceil(ThumbnailCount / (TileWidth * TileHeight))`. Les planches
//! passent par la route trickplay DÉDIÉE du backend
//! (`/api/jellyfin/items/{id}/trickplay/{width}/{index}.jpg`), enregistrées
//! sous `meta/<item>/trickplay/<width>/<index>.jpg` + un résumé `trickplay.json`.

use super::fsops;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::io::Read;
use std::path::Path;

/// Sérialisé en PascalCase pour coller au type partagé `TrickplayInfo`
/// (packages/shared) — le lecteur le consomme tel quel, sans conversion.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct TrickplayInfo {
    pub width: i64,
    pub height: i64,
    pub tile_width: i64,
    pub tile_height: i64,
    pub thumbnail_count: i64,
    pub interval: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalTrickplay {
    pub media_source_id: String,
    pub width: i64,
    pub info: TrickplayInfo,
}

fn info_from_value(v: &Value) -> Option<TrickplayInfo> {
    let get = |k: &str| v.get(k).and_then(Value::as_i64);
    Some(TrickplayInfo {
        width: get("Width")?,
        height: get("Height")?,
        tile_width: get("TileWidth").filter(|n| *n > 0)?,
        tile_height: get("TileHeight").filter(|n| *n > 0)?,
        thumbnail_count: get("ThumbnailCount").filter(|n| *n > 0)?,
        interval: get("Interval").filter(|n| *n > 0)?,
    })
}

/// Largeur préférée (320 sinon la plus proche) pour le mediaSourceId choisi.
fn pick_width(manifest: &Value, media_source_id: &str) -> Option<(String, i64, TrickplayInfo)> {
    let by_source = manifest.as_object()?;
    let (msrc, widths) = by_source
        .get_key_value(media_source_id)
        .or_else(|| by_source.iter().next())?;
    let widths = widths.as_object()?;
    let mut best: Option<(i64, TrickplayInfo)> = None;
    for (w, info_val) in widths {
        let Ok(width) = w.parse::<i64>() else { continue };
        let Some(info) = info_from_value(info_val) else { continue };
        let better = match &best {
            None => true,
            Some((bw, _)) => (width - 320).abs() < (bw - 320).abs(),
        };
        if better {
            best = Some((width, info));
        }
    }
    best.map(|(w, info)| (msrc.clone(), w, info))
}

pub fn tile_count(info: &TrickplayInfo) -> i64 {
    let per_tile = info.tile_width * info.tile_height;
    if per_tile <= 0 {
        return 0;
    }
    (info.thumbnail_count + per_tile - 1) / per_tile
}

fn fetch_tile(agent: &ureq::Agent, url: &str, token: &str) -> Option<Vec<u8>> {
    let response = agent.get(url).set("X-Emby-Token", token).call().ok()?;
    let mut bytes = Vec::new();
    response
        .into_reader()
        .take(6 * 1024 * 1024)
        .read_to_end(&mut bytes)
        .ok()?;
    if bytes.is_empty() {
        None
    } else {
        Some(bytes)
    }
}

/// Télécharge le manifeste + toutes les planches. `item_json` = octets de
/// l'item.json déjà récupéré (avec fields=Trickplay). Retourne le nombre de
/// planches enregistrées (0 si pas de trickplay). Best-effort.
pub fn download(
    agent: &ureq::Agent,
    server_url: &str,
    token: &str,
    root: &Path,
    item_id: &str,
    media_source_id: &str,
    item_json: &[u8],
) -> usize {
    let Ok(parsed) = serde_json::from_slice::<Value>(item_json) else {
        return 0;
    };
    let Some(manifest) = parsed.get("Trickplay") else { return 0 };
    if manifest.is_null() {
        return 0;
    }
    let Some((msrc, width, info)) = pick_width(manifest, media_source_id) else {
        return 0;
    };
    let count = tile_count(&info);
    if count <= 0 || count > 10_000 {
        return 0;
    }

    let dir = format!("meta/{item_id}/trickplay/{width}");
    let mut saved = 0;
    for index in 0..count {
        let rel = format!("{dir}/{index}.jpg");
        let Ok(path) = fsops::safe_join(root, &rel) else { continue };
        if path.exists() {
            saved += 1;
            continue;
        }
        let url = format!(
            "{server_url}/api/jellyfin/items/{item_id}/trickplay/{width}/{index}.jpg?mediaSourceId={msrc}"
        );
        let Some(bytes) = fetch_tile(agent, &url, token) else { continue };
        if let Some(parent) = path.parent() {
            if std::fs::create_dir_all(parent).is_err() {
                continue;
            }
        }
        if std::fs::write(&path, &bytes).is_ok() {
            saved += 1;
        }
    }

    // Résumé pour la résolution côté lecteur (meta/<item>/trickplay.json).
    if saved > 0 {
        let summary = LocalTrickplay { media_source_id: msrc, width, info };
        if let Ok(json) = serde_json::to_vec(&summary) {
            if let Ok(path) = fsops::safe_join(root, &format!("meta/{item_id}/trickplay.json")) {
                let _ = std::fs::write(path, json);
            }
        }
    }
    saved as usize
}

/// Le manifeste trickplay local est-il déjà présent ?
pub fn exists(root: &Path, item_id: &str) -> bool {
    fsops::safe_join(root, &format!("meta/{item_id}/trickplay.json"))
        .map(|p| p.exists())
        .unwrap_or(false)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn info() -> TrickplayInfo {
        TrickplayInfo {
            width: 320, height: 180, tile_width: 10, tile_height: 10,
            thumbnail_count: 720, interval: 10_000,
        }
    }

    #[test]
    fn nombre_de_planches_arrondi_au_superieur() {
        assert_eq!(tile_count(&info()), 8); // ceil(720 / 100)
        assert_eq!(tile_count(&TrickplayInfo { thumbnail_count: 100, ..info() }), 1);
        assert_eq!(tile_count(&TrickplayInfo { thumbnail_count: 101, ..info() }), 2);
    }

    #[test]
    fn choix_de_largeur_prefere_320() {
        let manifest = serde_json::json!({
            "msrc-A": {
                "160": {"Width":160,"Height":90,"TileWidth":10,"TileHeight":10,"ThumbnailCount":720,"Interval":10000},
                "320": {"Width":320,"Height":180,"TileWidth":10,"TileHeight":10,"ThumbnailCount":720,"Interval":10000}
            }
        });
        let (msrc, width, _) = pick_width(&manifest, "msrc-A").unwrap();
        assert_eq!(msrc, "msrc-A");
        assert_eq!(width, 320);
        // mediaSourceId absent → repli sur le premier.
        assert_eq!(pick_width(&manifest, "inconnu").unwrap().0, "msrc-A");
    }

    #[test]
    fn download_sans_trickplay_renvoie_zero() {
        let tmp = tempfile::tempdir().unwrap();
        let agent = ureq::AgentBuilder::new().build();
        let n = download(&agent, "http://127.0.0.1:1", "tok", tmp.path(), "item1", "ms1", b"{}");
        assert_eq!(n, 0);
    }
}

File: apps/desktop-electron/src/main/downloads/trickplay.ts
/**
 * Tuiles trickplay — l'aperçu au survol de la barre de progression.
 *
 * Manifeste : `item.Trickplay[mediaSourceId][width]`, qui exige
 * `fields=Trickplay` sur l'item. Nombre de planches, vérifié dans la source de
 * Jellyfin v10.11 : `ceil(ThumbnailCount / (TileWidth × TileHeight))`.
 *
 * Les planches passent par la route trickplay DÉDIÉE du backend et sont
 * enregistrées sous `meta/<item>/trickplay/<width>/<index>.jpg`, avec un
 * résumé `trickplay.json` que le lecteur consomme tel quel.
 *
 * Portage de `apps/desktop/src-tauri/src/downloads/trickplay.rs`.
 */

import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import path from "node:path";
import { MAX_TILE_BYTES, type FetchBytes } from "./fetcher";
import { asInteger, asRecord, field } from "./json";
import { mediaFileExists, safeJoin } from "./paths";

/**
 * Sérialisé en PascalCase pour coller au type partagé `TrickplayInfo`
 * (`packages/shared`) — le lecteur le consomme sans conversion.
 */
export interface TrickplayInfo {
  Width: number;
  Height: number;
  TileWidth: number;
  TileHeight: number;
  ThumbnailCount: number;
  Interval: number;
}

interface LocalTrickplay {
  mediaSourceId: string;
  width: number;
  info: TrickplayInfo;
}

/** Largeur visée : l'aperçu fait environ 320 px de large dans l'interface. */
const TARGET_WIDTH = 320;

/** Garde-fou : au-delà, le manifeste ment ou l'item est aberrant. */
const MAX_SHEETS = 10_000;

/**
 * Marqueur « le serveur n'a pas de trickplay pour cet item ».
 *
 * Sans lui, la réparation redemandait `Items/<id>?fields=Trickplay` pour CHAQUE
 * item complet, à CHAQUE démarrage : un film dont la bibliothèque ne génère pas
 * de planches n'en aura jamais, et la requête repartait à vie. Une bibliothèque
 * peut cependant se mettre à en produire — le marqueur périme donc.
 *
 * Un fichier plutôt qu'une colonne : la base est PARTAGÉE avec l'app Tauri, et
 * monter `PRAGMA user_version` la rendrait illisible pour elle. Ce fichier-ci
 * lui est simplement invisible : elle ne lit que des noms précis.
 */
const MARKER = "trickplay.none";

/** Au-delà, on redemande au serveur. */
export const RECHECK_AFTER_MS = 30 * 24 * 60 * 60 * 1000;

function infoFromValue(value: unknown): TrickplayInfo | null {
  const positive = (key: string): number | null => {
    const n = asInteger(field(value, key));
    return n !== null && n > 0 ? n : null;
  };
  const width = asInteger(field(value, "Width"));
  const height = asInteger(field(value, "Height"));
  const tileWidth = positive("TileWidth");
  const tileHeight = positive("TileHeight");
  const thumbnailCount = positive("ThumbnailCount");
  const interval = positive("Interval");
  if (
    width === null ||
    height === null ||
    tileWidth === null ||
    tileHeight === null ||
    thumbnailCount === null ||
    interval === null
  ) {
    return null;
  }
  return {
    Width: width,
    Height: height,
    TileWidth: tileWidth,
    TileHeight: tileHeight,
    ThumbnailCount: thumbnailCount,
    Interval: interval,
  };
}

/** Largeur la plus proche de la cible, pour la source demandée. */
export function pickWidth(
  manifest: unknown,
  mediaSourceId: string,
): { mediaSourceId: string; width: number; info: TrickplayInfo } | null {
  const bySource = asRecord(manifest);
  if (bySource === null) return null;

  // La source demandée d'abord ; à défaut la première — un manifeste ne porte
  // qu'une source dans l'immense majorité des cas.
  const key = mediaSourceId in bySource ? mediaSourceId : Object.keys(bySource)[0];
  if (key === undefined) return null;
  const widths = asRecord(bySource[key]);
  if (widths === null) return null;

  let best: { width: number; info: TrickplayInfo } | null = null;
  for (const [raw, value] of Object.entries(widths)) {
    const width = Number.parseInt(raw, 10);
    if (!Number.isFinite(width)) continue;
    const info = infoFromValue(value);
    if (info === null) continue;
    const closest =
      best === null ||
      Math.abs(width - TARGET_WIDTH) < Math.abs(best.width - TARGET_WIDTH);
    if (closest) best = { width, info };
  }
  return best === null ? null : { mediaSourceId: key, ...best };
}

/** Nombre de planches, arrondi AU SUPÉRIEUR. */
export function tileCount(info: TrickplayInfo): number {
  const perSheet = info.TileWidth * info.TileHeight;
  if (perSheet <= 0) return 0;
  return Math.ceil(info.ThumbnailCount / perSheet);
}

/**
 * Télécharge le manifeste puis toutes les planches manquantes.
 *
 * `itemJson` = octets de l'`item.json` déjà récupéré, avec `fields=Trickplay`.
 * Retourne le nombre de planches disponibles à la fin. Best-effort et
 * idempotent : ce qui est déjà là est compté sans être retéléchargé.
 *
 * Les trois sorties précoces posent le marqueur `trickplay.none` : à ce
 * stade la source a répondu, et sa réponse dit qu'il n'y a rien à prendre.
 * Un échec de planche, lui, ne le pose pas — c'est du réseau.
 */
export async function download(
  fetchBytes: FetchBytes,
  serverUrl: string,
  root: string,
  itemId: string,
  mediaSourceId: string,
  itemJson: unknown,
  nowMs: number,
): Promise<number> {
  const manifest = field(itemJson, "Trickplay");
  if (manifest === null || manifest === undefined) {
    markNone(root, itemId, nowMs);
    return 0;
  }

  const choice = pickWidth(manifest, mediaSourceId);
  if (choice === null) {
    markNone(root, itemId, nowMs);
    return 0;
  }
  const sheets = tileCount(choice.info);
  if (sheets <= 0 || sheets > MAX_SHEETS) {
    markNone(root, itemId, nowMs);
    return 0;
  }

  const folder = `meta/${itemId}/trickplay/${choice.width}`;
  let fetched = 0;
  for (let index = 0; index < sheets; index += 1) {
    let target: string;
    try {
      target = safeJoin(root, `${folder}/${index}.jpg`);
    } catch {
      continue;
    }
    if (existsSync(target)) {
      fetched += 1;
      continue;
    }

    const url =
      `${serverUrl}/api/jellyfin/items/${itemId}/trickplay/${choice.width}/${index}.jpg` +
      `?mediaSourceId=${choice.mediaSourceId}`;
    const bytes = await fetchBytes(url, MAX_TILE_BYTES);
    if (bytes === null || bytes.byteLength === 0) continue;

    try {
      mkdirSync(path.dirname(target), { recursive: true });
      writeFileSync(target, bytes);
      fetched += 1;
    } catch {
      // Une planche manquante dégrade l'aperçu, elle ne casse rien.
    }
  }

  if (fetched > 0) {
    // Un marqueur d'un passage précédent n'a plus lieu d'être : la
    // bibliothèque s'est mise à générer des planches depuis.
    forgetMarker(root, itemId);
    const resume: LocalTrickplay = {
      mediaSourceId: choice.mediaSourceId,
      width: choice.width,
      info: choice.info,
    };
    try {
      writeFileSync(safeJoin(root, `meta/${itemId}/trickplay.json`), JSON.stringify(resume));
    } catch {
      // Sans résumé, le lecteur retombe sur l'aperçu serveur : pas bloquant.
    }
  }
  return fetched;
}

/** Le manifeste trickplay local est-il déjà là ? */
export function exists(root: string, itemId: string): boolean {
  return mediaFileExists(root, `meta/${itemId}/trickplay.json`);
}

function markerPath(root: string, itemId: string): string | null {
  try {
    return safeJoin(root, `meta/${itemId}/${MARKER}`);
  } catch {
    return null;
  }
}

/**
 * Note que le serveur ne propose PAS de trickplay pour cet item.
 *
 * Posé uniquement quand la source a répondu et n'a rien à offrir — jamais sur
 * un échec réseau, sans quoi une coupure passagère priverait l'item d'aperçu
 * pour un mois.
 */
export function markNone(root: string, itemId: string, nowMs: number): void {
  const target = markerPath(root, itemId);
  if (target === null) return;
  try {
    mkdirSync(path.dirname(target), { recursive: true });
    writeFileSync(target, String(nowMs));
  } catch {
    // Sans marqueur, on redemandera : c'est le comportement d'avant, pas une panne.
  }
}

/** Item constaté sans trickplay, et depuis moins d'un mois ? */
export function noneRecently(root: string, itemId: string, nowMs: number): boolean {
  const target = markerPath(root, itemId);
  if (target === null) return false;
  try {
    const writtenAt = Number.parseInt(readFileSync(target, "utf8"), 10);
    return Number.isFinite(writtenAt) && nowMs - writtenAt < RECHECK_AFTER_MS;
  } catch {
    return false;
  }
}

function forgetMarker(root: string, itemId: string): void {
  const target = markerPath(root, itemId);
  if (target === null) return;
  try {
    rmSync(target, { force: true });
  } catch {
    // Un marqueur qui traîne à côté d'un trickplay.json présent est inerte :
    // `exists()` est consulté en premier.
  }
}
